8283660: Convert com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java finalizer to Cleaner

bchristi-git opened the pull request:
Please review this change to replace the finalizer in `AbstractLdapNamingEnumeration` with a Cleaner.

The pieces of state required for cleanup (`LdapCtx homeCtx`, `LdapResult res`, and `LdapClient enumClnt`) are moved to a static inner class . From there, the change is fairly mechanical.

Details of note: 
1. Some operations need to change the state values (the update() method is probably the most interesting).
2. Subclasses need to access `homeCtx`; I added a `homeCtx()` method to read `homeCtx` from the superclass's `state`.

The test case is based on a copy of `com/sun/jndi/ldap/blits/AddTests/AddNewEntry.java`. A more minimal test case might be possible, but this was done for expediency.

The test only confirms that the new Cleaner use does not keep the object reachable. It only tests `LdapSearchEnumeration` (not `LdapNamingEnumeration` or `LdapBindingEnumeration`, though all are subclasses of `AbstractLdapNamingEnumeration`). 

Thanks.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blocker
&nbsp;⚠️ Whitespace errors (failed with the updated jcheck configuration)

### Issue
 * [JDK-8283660](https://bugs.openjdk.org/browse/JDK-8283660): Convert com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java finalizer to Cleaner (**Sub-task** - P4)


### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [c8bdac44](https://git.openjdk.org/jdk/pull/8311/files/c8bdac443da1293bf2c437cc9946271556502341)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/8311/head:pull/8311` \
`$ git checkout pull/8311`

Update a local copy of the PR: \
`$ git checkout pull/8311` \
`$ git pull https://git.openjdk.org/jdk.git pull/8311/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8311`

View PR using the GUI difftool: \
`$ git pr show -t 8311`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8311.diff">https://git.openjdk.org/jdk/pull/8311.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/8311#issuecomment-1103356167)

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -42,17 +45,48 @@
+    /* This class maintains the pieces of state that need (or are needed for)
+     * cleanup, which happens by calling cleanup(), or is done by the Cleaner.
+     */
+    private static class CleaningAction implements Runnable {
```
This nested class might be more aptly named  `EnumCtx` since it is used during the enumeration process.

The mention of the `cleanup()` method in the nested class javadoc is a bit ambiguous, it seems there should be a cleanup() method in the class, but it is in AbstractLdapNamingEnumeration.

The `state` field might also be renamed `enumCtx`.

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -94,10 +127,10 @@
-            this.homeCtx = homeCtx;
-            homeCtx.incEnumCount();
-            enumClnt = homeCtx.clnt; // remember
+            this.state.homeCtx.incEnumCount();
```
Readability might be improved by using the new `homeCtx()` method instead of `state.homeCtx` in this file.
It would then be consistent with how subclasses access it.
It depends which style you prefer to be more consistent with.

dfuch reviewed (commented):
I also agree with Roger's suggestions.

bchristi-git commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -42,17 +45,48 @@
+        }
+    }
+
+    private CleaningAction state;
```
Makes sense to me. `cleanable` can be final, too.

bchristi-git commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -42,17 +45,48 @@
+        @Override
+        public void run() {
+            if (enumClnt != null) {
+                enumClnt.clearSearchReply(res, homeCtx.reqCtls);
```
Yes, it is strange -- that code came from the finalizer. I will add an assert.

bchristi-git replied:
It appears that the update() method can set 'homeCtx' for 'ne' to null while leaving 'enumClnt' non-null (~L410).
Perhaps here, the clearSearchReply() call should only happen if both are non-null.

bchristi-git commented:
`AbstractLdapEnumeration`'s mutable state brings the possibility of threading issues between the program and cleaner threads. I've added some threading-related changes to the fix. See my [comment in the bug report](https://bugs.openjdk.java.net/browse/JDK-8283660?focusedCommentId=14498566&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14498566) for additional background.

Since synchronization may now happen on the cleaner thread, I've changed `AbstractLdapEnumeration` to use its own `Cleaner` instance instead of the shared cleaner, for added safety. There have been deadlocks in ldap cleanup in the past.

The added `finally` blocks led to a lot of indentation changes. The "hide whitespace" option might help with viewing the changes.

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -126,45 +170,50 @@ public final boolean hasMoreElements() {
+        try {
+            LdapResult newRes = homeCtx().getSearchReply(enumCtx.enumClnt, enumCtx.res);
+            enumCtx.setRes(newRes);
+            if (enumCtx.res == null) {
```
This looks odd, setting the value using synchronized, but reading it without.

bchristi-git replied:
I've added getters to EnumCtx, and a comment explaining why only setters are synchronized.

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -126,45 +170,50 @@ public final boolean hasMoreElements() {
+            }
+        } finally {
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(this);
```
I don't think there is any benefit to the `try{} finally {fence}`.
The reachabilityFence has no executable code. Its only purpose is to keep the reference in scope alive.

plevart replied:
> I don't think there is any benefit to the `try{} finally {fence}`. The reachabilityFence has no executable code. Its only purpose is to keep the reference in scope alive.

...reachability fence has no code, but puting it into finally block makes the compiler see it on all possible exit paths from the try block (any possible exceptions, returns, etc.) so the argument 'this' is protected from being GC-ed until every possible code in try block is finished.

plevart replied:
The memory fence is not needed though (as explained above).

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -362,45 +415,37 @@ protected final boolean hasMoreReferrals() throws NamingException {
+            refEx = ne.refEx;
+            listArg = ne.listArg;
+        } finally {
+            // Ensure Cleaner does not run until after this method completes
```
Ditto try/finally is unnecessary.

plevart replied:
> Ditto try/finally is unnecessary.

Perhaps, because there may be only one exit path. But style-wise it is still better to play safe for any possible future changes to this part of code. Memory fence is not needed though as explained above.

RogerRiggs commented:
Hans, thank for the detailed example. I had not fully considered the flow of control in the throwing case.

bchristi-git commented:
I've updated the test case to confirm that cleanup is performed correctly, by inspecting the enumCount within the LdapCtx.

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -94,10 +146,10 @@
-            homeCtx.incEnumCount();
-            enumClnt = homeCtx.clnt; // remember
+            this.enumCtx.homeCtx.incEnumCount();
+            this.cleanable = LDAP_CLEANER.register(this, enumCtx);
```
Use  `enumCtx.getHomeCts()` above.
Use `this.enumCtx` in call to `register`.
For consistency.

RogerRiggs reviewed (approved):
trivial comments.
The commented out printf/println's should be removed before committing.

bchristi-git commented:
> The commented out printf/println's should be removed before committing.

Do you mean the pre-existing `println`s in LdapSearchEnumeration.java?

bchristi-git commented:
I've updated how the reachability and memory visibility issues (per [comment](https://bugs.openjdk.org/browse/JDK-8283660?focusedCommentId=14498566&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14498566)) are addressed:

**Additional reachability fences**

Unless it's _immediately obvious_ that _**no**_ cleanable state is **_accessed_** during the course of a method's execution, all methods should include a `reachabilityFence` to ensure that cleanable state is not cleaned up while a method is still running. This applies to any class that uses a `Cleaner` (or a finalizer, really).

**`VarHandle.fullFence` to ensure memory visibility, instead of synchronized methods**
This makes for simpler code, and better reflects our intentions.

Unless it's _immediately obvious_ that **_no_** cleanable state is **_modified_** during the course of the method's execution, all methods should include a `fullFence` to ensure changes made on the main/program thread are seen by the cleanup thread. This applies to any class that uses a `Cleaner` (or a finalizer, really) where the state to be cleaned can be mutated during the course of execution.

Other possibilities for triggering the necessary volatile operations could be:
* empty synchronized blocks
* adding some "junk" variable to EnumCtx, along with volatile reads/writes of the variable where needed
---
One could perhaps use deep code tracing to determine that cleanable state is not accessed/written during the course of a method's execution, and omit fences based on that knowledge. However I believe that leaves too large a burden on future maintainers and reviewers to remember to re-perform the code tracing and re-confirm non-use of cleanable state, even when changing possibly "far away" code.

RogerRiggs commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -364,49 +493,43 @@ protected final boolean hasMoreReferrals() throws NamingException {
+            entries = ne.entries;
+            refEx = ne.refEx;
+            listArg = ne.listArg;
+			// record a previous exception and quit if any limit is reached
```
Replace tabs with spaces.

RogerRiggs reviewed (commented):
LGTM 
(Except for a TODO and some tabs that should be spaces)
(Probably there's a higher level lesson to be learned about designing this kind of interaction with a server and the teardown needed).

stuart-marks commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -94,16 +141,38 @@
+            // are reachability fences to ensure that the registered object remains
+            // reachable.
+            // TODO: Is anything else needed so that this constructor
+            //  "happens-before" the cleaning action ?
```
I think the resolution of this TODO is to modify the specification of Cleaner to require that:

1. the object being registered is guaranteed reachable until after registration has completed; and
2. a "memory consistency effects" clause should also be added to Cleaner.register. For an example, see the class specification of CopyOnWriteArrayList.

It may be that reachabilityFences are already in the right place in Cleaner. Great. But last time I looked it doesn't have any memory visibility operations. I'd suggest opening a bug on Cleaner for the spec change and accumulating notes there.

stuart-marks commented:
I think the main outstanding issues here are as follows.

First, are we convinced that it's necessary to add a try/finally block with a memory fence and a reachability fence in pretty much every mutator method? I think the answer is Yes, but we ought to have this ratified by memory model and GC experts.

A second issue is whether VarHandle::fullFence is the right memory fence operation. I'm not quite convinced that it is, but I'm not sure what the right alternative is either.

Finally, we probably also need approval from the maintainers of this code. :-)

stuart-marks commented:
Hans Boehm wrote:

> I also have concerns about the use of fullFence here. I believe it should be the case that reachabilityFence guarantees visibility of memory operations program-ordered before the reachabilityFence(p) to the Cleaner associated with p. Mutator-collector synchronization should normally ensure that. On rereading, it also appears to me that the current reachabilityFence() documentation does not make that as clear as it should.

> It appears to me that this is addressing an instance of the problem for which I suggested a more general, though also not completely elegant, solution in https://docs.google.com/document/d/1yMC4VzZobMQTrVAraMy7xBdWPCJ0p7G5r_43yaqsj-s/edit?usp=sharing .

Hi Hans, thanks for looking at this. In the prior discussions on reachabilityFence and premature finalization, I don't recall seeing any mention of memory model issues. To my mind the discussions here are the first mention of them. (Of course it's possible I could have missed something.) The memory model is quite subtle and it's difficult to be sure of anything. However, as time goes on we've become more confident that there _IS_ an issue here with the memory model that needs to be addressed. 

Then there is a the question of what to do about it. One possibility is to add some memory ordering semantics into reachabilityFence(p), as you suggest. As convenient as it might seem, it's not obvious to me that we want reachability fused with memory order. At least we should discuss them separately before deciding what to do.

This PR seems to be on a long journey :-) so let me describe some of the background and where I think this ought to go.

First, like most PRs, this started off as an effort to make some code changes. In this case it's part of Brent's (@bchristi-git) effort to convert finalizers in the JDK to Cleaners. This is related to discovering coding patterns for how to do this effectively. For example, the entire object's state is available to the finalizer. But in order to convert this to a Cleaner, the state available to the cleaning action needs to be refactored into a separate object from the "outer" object. The `EnumCtx` nested class serves this role here.

Second, we're trying to address the reachability issues. You (Hans) have been writing and speaking about this for some time, mostly in the context of finalization. We in the JDK group haven't prioritized fixing this issue with respect to finalization (since it's old and deprecated and nobody should be using it, yeah right). Now that we're converting things to use Cleaner, which has the same issues, we're forced to confront them. Our working position is that there needs to be a reachabilityFence within a try/finally block in the "right" places; determining the definition of "right" is one of the goals here.

The third issue is memory ordering. For finalization-based objects, the JLS specifies a happens-before edge between the constructor and the finalizer. (I think this works for objects whose finalizable state is fully initialized in the constructor. But it doesn't work for objects, like this one, whose finalizable state is mutable throughout the lifetime of the object.) There is nothing specified for memory ordering edges for Cleaner or java.lang.ref references at all that I can see. Given the lack of such specifications, we're faced with using the right low-level memory ordering mechanisms to get the memory order we require. We're using VarHandle::fullFence as kind of a placeholder for this. (We've also considered empty synchronized blocks, writes/reads to "junk" variables created expressly for this purpose, and other VarHandle fence operations, but none seem obviously better. I'm sure there are other alternatives we haven't considered.) I'd welcome discussion of the proper alternative.

The fourth issue is, do we really want every programmer who uses Cleaner to have to figure out all this reachabilityFence and VarHandle fence stuff? Of course not. It would be nice to have some higher-level construct (such as a language change like the "Reachability Revisited" proposal), or possibly to create some library APIs to facilitate this. At a minimum, I think we need to adjust various specifications like Cleaner and java.lang.ref to address memory ordering issues. There is certainly more that needs to be done though.

The difficulty with trying to come up with language changes or library APIs is that I don't think we understand this problem well enough to define what those mechanisms are actually supposed to do. So before we get to that point, I think we should see attempt to write down a correct solution using the existing reachability and memory ordering mechanisms. That's what Brent has done here. We should review and iterate on this and identify the places where the specs need to change, and arrive at a point where the we believe the code, even if ugly and inconvenient, is correct under that spec.

Maybe at that point we can contemplate a higher-level approach such as a language mechanism or a library API (or both), but I don't think we're quite there yet. Or maybe some alternative path forward might emerge.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -94,16 +141,38 @@
+                // overridden to access the cleanable state without the proper fences.
+
+                // Ensure writes are visible to the Cleaner thread
+                VarHandle.fullFence();
```
I don't think any memory fences are needed here. Reachability fence is enough. Cleaner runs in its own thread by polling enqued PhantomReference(s) from the ReferenceQueue. There is a happens-before edge between ReferenceHandler thread enqueue-ing "pending" PhantomReferences and Cleaner thread dequeue-ing them. There is a happens-before edge between GC thread clearing a PhantomReference and linking it into a "pending" list and ReferenceHandler thread unlinking it from the "pending" list and enque-ing it into a ReferenceQueue. There is a happens-before edge before a call to Reference.reachabilityFence(referent) and a GC thread discovering a phantom-reachable referent and clearing the PhantomReference and linking it into a "pending" list.
So there you have a happens-before chain which makes all writes before a call to eference.reachabilityFence(this) visible to a Cleaner task that is initialized to observe reachabillity of this....

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -114,7 +183,16 @@ public final T nextElement() {
+                // Same situation as nextElement() - see comment above
+
+                // Ensure writes are visible to the Cleaner thread
+                VarHandle.fullFence();
```
Same as above. Memory fences are not needed.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -176,17 +260,23 @@ private void getNextBatch() throws NamingException {
+            }
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
The memory fence is not needed (as explained above), but...

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -176,17 +260,23 @@ private void getNextBatch() throws NamingException {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```
The reachability fence should take 'this' as a parameter, not enumCtx. Since 'this' reachability is tracked by Cleaner and not 'enumCtx' reachability.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -195,49 +285,62 @@ public final boolean hasMore() throws NamingException {
+            return nextImpl();
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
The memory fence is not needed (as explained above), but...

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -195,49 +285,62 @@ public final boolean hasMore() throws NamingException {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```
The reachability fence should take 'this' as a parameter, not enumCtx. Since 'this' reachability is tracked by Cleaner and not 'enumCtx' reachability.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -195,49 +285,62 @@ public final boolean hasMore() throws NamingException {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```
The same as above: no need for memory fence and reachability fence should take 'this' as parameter.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -251,22 +354,33 @@ private T nextImpl() throws NamingException {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```
Same as above: no need for memory fence and reachability fence should take 'this' as parameter.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -289,15 +404,23 @@ protected abstract T createItem(String dn, Attributes attrs,
+            }
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
No need for memory fence.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -309,51 +358,55 @@ protected abstract AbstractLdapNamingEnumeration<? extends NameClassPair> getRef
-            return (false);
+        } finally {
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```
> Ditto, the try/finally is unnecessary.

I think it is necessary for reachability fence since try block has many exit paths and accesses state cleaned up by Cleaner. Memory fence OTOH is not needed.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapBindingEnumeration.java:
```diff
@@ -54,57 +56,71 @@
+            return binding;
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
no need for memory fence.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapBindingEnumeration.java:
```diff
@@ -54,57 +56,71 @@
+            return (AbstractLdapNamingEnumeration<? extends NameClassPair>) refCtx.listBindings(listArg);
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
no need for memory fence.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapNamingEnumeration.java:
```diff
@@ -46,35 +47,49 @@
+            return ncp;
+        }  finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
no need for memory fence.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapNamingEnumeration.java:
```diff
@@ -46,35 +47,49 @@
+            return (AbstractLdapNamingEnumeration<? extends NameClassPair>)refCtx.list(listArg);
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
no need for memory fence

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapSearchEnumeration.java:
```diff
@@ -58,134 +60,145 @@
+            return sr;
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
memory fence is not needed.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapSearchEnumeration.java:
```diff
@@ -194,14 +207,24 @@ public void appendUnprocessedReferrals(LdapReferralException ex) {
+                    searchArgs.name, searchArgs.filter, searchArgs.cons);
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
```
no need for memory fence.

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -42,17 +47,60 @@
+        @Override
+        public void run() {
+            // Ensure changes on the main/program thread happens-before cleanup
+            VarHandle.fullFence();
```
no need for memory fence as explained in various finally blocks...

plevart commented on src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java:
```diff
@@ -42,17 +47,60 @@
+            if (homeCtx != null) {
+                homeCtx.decEnumCount();
+                homeCtx = null;
+            }
```
Cleaner's task (run() method) is guaranteed to be called at most once. So there's no need for above "idempotent" logic. Unless some of those fields are optional when the instance is constructed....

plevart commented:
> Hans Boehm wrote:
> 
> > I also have concerns about the use of fullFence here. I believe it should be the case that reachabilityFence guarantees visibility of memory operations program-ordered before the reachabilityFence(p) to the Cleaner associated with p. Mutator-collector synchronization should normally ensure that. On rereading, it also appears to me that the current reachabilityFence() documentation does not make that as clear as it should.
> 
> > It appears to me that this is addressing an instance of the problem for which I suggested a more general, though also not completely elegant, solution in https://docs.google.com/document/d/1yMC4VzZobMQTrVAraMy7xBdWPCJ0p7G5r_43yaqsj-s/edit?usp=sharing .
> 
> Hi Hans, thanks for looking at this. In the prior discussions on reachabilityFence and premature finalization, I don't recall seeing any mention of memory model issues. To my mind the discussions here are the first mention of them. (Of course it's possible I could have missed something.) The memory model is quite subtle and it's difficult to be sure of anything. However, as time goes on we've become more confident that there _IS_ an issue here with the memory model that needs to be addressed.
> 
> Then there is a the question of what to do about it. One possibility is to add some memory ordering semantics into reachabilityFence(p), as you suggest. As convenient as it might seem, it's not obvious to me that we want reachability fused with memory order. At least we should discuss them separately before deciding what to do.
> 
> This PR seems to be on a long journey :-) so let me describe some of the background and where I think this ought to go.
> 
> First, like most PRs, this started off as an effort to make some code changes. In this case it's part of Brent's (@bchristi-git) effort to convert finalizers in the JDK to Cleaners. This is related to discovering coding patterns for how to do this effectively. For example, the entire object's state is available to the finalizer. But in order to convert this to a Cleaner, the state available to the cleaning action needs to be refactored into a separate object from the "outer" object. The `EnumCtx` nested class serves this role here.
> 
> Second, we're trying to address the reachability issues. You (Hans) have been writing and speaking about this for some time, mostly in the context of finalization. We in the JDK group haven't prioritized fixing this issue with respect to finalization (since it's old and deprecated and nobody should be using it, yeah right). Now that we're converting things to use Cleaner, which has the same issues, we're forced to confront them. Our working position is that there needs to be a reachabilityFence within a try/finally block in the "right" places; determining the definition of "right" is one of the goals here.
> 
> The third issue is memory ordering. For finalization-based objects, the JLS specifies a happens-before edge between the constructor and the finalizer. (I think this works for objects whose finalizable state is fully initialized in the constructor. But it doesn't work for objects, like this one, whose finalizable state is mutable throughout the lifetime of the object.) There is nothing specified for memory ordering edges for Cleaner or java.lang.ref references at all that I can see. Given the lack of such specifications, we're faced with using the right low-level memory ordering mechanisms to get the memory order we require. We're using VarHandle::fullFence as kind of a placeholder for this. (We've also considered empty synchronized blocks, writes/reads to "junk" variables created expressly for this purpose, and other VarHandle fence operations, but none seem obviously better. I'm sure there are other alternatives we haven't considered.) I'd welcome discussion of the proper alternative.
> 
> The fourth issue is, do we really want every programmer who uses Cleaner to have to figure out all this reachabilityFence and VarHandle fence stuff? Of course not. It would be nice to have some higher-level construct (such as a language change like the "Reachability Revisited" proposal), or possibly to create some library APIs to facilitate this. At a minimum, I think we need to adjust various specifications like Cleaner and java.lang.ref to address memory ordering issues. There is certainly more that needs to be done though.
> 
> The difficulty with trying to come up with language changes or library APIs is that I don't think we understand this problem well enough to define what those mechanisms are actually supposed to do. So before we get to that point, I think we should see attempt to write down a correct solution using the existing reachability and memory ordering mechanisms. That's what Brent has done here. We should review and iterate on this and identify the places where the specs need to change, and arrive at a point where the we believe the code, even if ugly and inconvenient, is correct under that spec.
> 
> Maybe at that point we can contemplate a higher-level approach such as a language mechanism or a library API (or both), but I don't think we're quite there yet. Or maybe some alternative path forward might emerge.

Hi,

Sorry for not reading up to this message before starting reviewing the changes... As I was trying to explain in the comments, I don't think there is any issue about memory ordering here. As Cleaner is implemented currently in OpenJDK, there a 5 threads involved in a typical scenario of a cleanable object:
T1 - initializing thread that registers a Cleaner action
T2 - some thread that accesses/mutates object state including parts that are accessed/cleaned up by Cleaner action and guards against premature Cleaner action execution with reachability fence
T3 - GC thread that discovers a phantom-reachable referent of a PhantomReference that is tracked by the Cleaner and clears the PhantomReference and links it into a "pending" list.
T4 - a ReferenceHandler thread that waits for "pending" Reference(s) and enqueues them to ReferenceQueue(s)
T5 - a single Cleaner thread that dequeues PhantomReference(s) from the Cleaner's ReferenceQueue and executes cleanup actions

There is a synchronization action between T1 and T5, so there is no doubt that writes in the construrctor of a tracked object that typically preced Cleaner registration are visible to Cleaner action.

There is a synchronization action between T3 and T4 and there is a synchronization action between T4 and T5 (both can be verified in code). The only question here remaining is whether there is synchronization between threads that mutate object state and GC thread that discovers phantom-reachable referents..

I'm not qualified to answer that question. Especially with the advent of new fully concurrent GC algorithms such as ZGC and Shenandoah. So perhaps this is a question for GC experts.

plevart commented:
...while waiting for a GC expert to confirm that writes, in program order preceding a call to reachability fence, are visibile to GC thread that discovers a phantom reachable referent, here is an example which illustrates that some writes at least "must" be visible for GC to function properly:

```java
var ar = new Object[1];
var e = new Object();

ar[0] = e;
Reference.reachabilityFence(e);

// GC kicks-in at this point

var e2 = ar[0];
// use e2
```

A write to an array slot ar[0] must be visible to a GC thread when it searches for root(s) after the mutator thread's call to reachability fence. If it was not, it could miss that fact that object 'e' is still reachable. So GC must do something so that at least writes to reference variables are visible. If it does not distinguish reference writes from primitive writes, then it does the same for all writes.

stuart-marks commented:
Hi Peter, thanks for contributing to this. I think you're observing that as a practical matter, the implementations of GC and of various libraries such as reference queues must see a consistent view of memory in order for anything to work. This seems right. However, I'm concerned about what the _specification_ says, or ought to say.

Hans seems to think that the JLS assertions about finalization also apply to reference processing. I'm not so sure... that seems to me to be a rather generous interpretation. On the face of it, I cannot find anything explicit in the specifications that supports it. It certainly seems reasonable (to me) that reachabilityFence(x) ought to HB the corresponding reference being dequeued. If so I would like the specification to say that. (The fact that there is a GC thread that enqueues the reference is part of the implementation, which is opaque to the specification.) It may also be the that any point where the referent is reachable HB its corresponding reference is dequeued. Clearly the GC implementation needs to make sure this is the case; the questions in my mind are whether, where, and how this should be specified!

It may be that the VarHandle fences aren't necessary. However, if they end up driving the right updates to the specifications, they will have served their purpose.

Setting this aside, it does seem like all uses of a cleanable object need to have a try/finally statement, with at least an RF in the finally clause. Is there any evidence that shows that this construct isn't needed?

bchristi-git commented:
Sorry for the delay. I'll try to get back to this soon.

turbanoff commented on src/java.naming/share/classes/com/sun/jndi/ldap/LdapNamingEnumeration.java:
```diff
@@ -46,35 +47,49 @@
+            }
+            ncp.setNameInNamespace(dn);
+            return ncp;
+        }  finally {
```
let's remove one redundant space after `}`

bchristi-git commented:
Not today, PullrequestCloserBot

bchristi-git commented:
Thanks for the reminder, bridgekeeper

bchristi-git commented:
Keepalive

stuart-marks commented:
`Reference.reachabilityFence(this)`

bchristi-git commented:
ever diligent,
"Are you still working on this?"
asks bridgekeeper bot

stuart-marks commented:
Oh darn, looks like `Reference.reachabilityFence(this)` is broken!!

stuart-marks commented:
`MethodHandles.lookup().findVarHandle(PR.class, "8311").acquireFence()`

bchristi-git commented:
[Presses Snooze]

bchristi-git commented:
Not yet

bchristi-git commented:
Let's hang on to this for a bit longer

bchristi-git commented:
July already!

bchristi-git commented:
Your persistence finally paid off, **bridgekeeper** - well played.

stuart-marks commented:
"Who reviewth the PR of Death
Must answer me these questions three
Ere the PR reopened be."
